# Chapter 3 – Classification (Heart Disease dataset)

Hands-On Machine Learning (Géron, 2nd ed.), Chapter 3, applied to the UCI Heart Disease data (`heart.csv`, 1025 rows, binary `target`).

In [ ]:
import pandas as pd
df=pd.read_csv("heart.csv")

In [ ]:
df.shape

(1025, 14)

In [ ]:
df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0
1,53,1,0,140,203,1,0,155,1,3.1,0,0,3,0
2,70,1,0,145,174,0,1,125,1,2.6,0,0,3,0
3,61,1,0,148,203,0,1,161,0,0.0,2,1,3,0
4,62,0,0,138,294,1,1,106,0,1.9,1,3,2,0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB


In [ ]:
df.describe()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
count,1025.000000,1025.000000,1025.000000,1025.000000,1025.00000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000
mean,54.434146,0.695610,0.942439,131.611707,246.00000,0.149268,0.529756,149.114146,0.336585,1.071512,1.385366,0.754146,2.323902,0.513171
std,9.072290,0.460373,1.029641,17.516718,51.59251,0.356527,0.527878,23.005724,0.472772,1.175053,0.617755,1.030798,0.620660,0.500070
min,29.000000,0.000000,0.000000,94.000000,126.00000,0.000000,0.000000,71.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,48.000000,0.000000,0.000000,120.000000,211.00000,0.000000,0.000000,132.000000,0.000000,0.000000,1.000000,0.000000,2.000000,0.000000
50%,56.000000,1.000000,1.000000,130.000000,240.00000,0.000000,1.000000,152.000000,0.000000,0.800000,1.000000,0.000000,2.000000,1.000000
75%,61.000000,1.000000,2.000000,140.000000,275.00000,0.000000,1.000000,166.000000,1.000000,1.800000,2.000000,1.000000,3.000000,1.000000
max,77.000000,1.000000,3.000000,200.000000,564.00000,1.000000,2.000000,202.000000,1.000000,6.200000,2.000000,4.000000,3.000000,1.000000


In [ ]:
df[["age", "trestbps", "chol", "thalach", "oldpeak", "ca"]].describe()

,age,trestbps,chol,thalach,oldpeak,ca
count,1025.000000,1025.000000,1025.00000,1025.000000,1025.000000,1025.000000
mean,54.434146,131.611707,246.00000,149.114146,1.071512,0.754146
std,9.072290,17.516718,51.59251,23.005724,1.175053,1.030798
min,29.000000,94.000000,126.00000,71.000000,0.000000,0.000000
25%,48.000000,120.000000,211.00000,132.000000,0.000000,0.000000
50%,56.000000,130.000000,240.00000,152.000000,0.800000,0.000000
75%,61.000000,140.000000,275.00000,166.000000,1.800000,1.000000
max,77.000000,200.000000,564.00000,202.000000,6.200000,4.000000


In [ ]:
df["target"].value_counts()

,count
target,
1,526
0,499


In [ ]:
df["target"].value_counts(normalize=True)

,proportion
target,
1,0.513171
0,0.486829


In [ ]:
df.duplicated().sum()

np.int64(723)

In [ ]:
df[df.duplicated()]

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
15,34,0,1,118,210,0,1,192,0,0.7,2,0,2,1
31,50,0,1,120,244,0,1,162,0,1.1,2,0,2,1
43,46,1,0,120,249,0,0,144,0,0.8,2,0,3,0
55,55,1,0,140,217,0,1,111,1,5.6,0,0,3,0
61,66,0,2,146,278,0,0,152,0,0.0,1,1,2,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1020,59,1,1,140,221,0,1,164,1,0.0,2,0,2,1
1021,60,1,0,125,258,0,0,141,1,2.8,1,1,3,0
1022,47,1,0,110,275,0,0,118,1,1.0,1,1,2,0
1023,50,0,0,110,254,0,0,159,0,0.0,2,0,2,1


In [ ]:
df.index

RangeIndex(start=0, stop=1025, step=1)

In [ ]:
df[df.duplicated(keep=False)].sort_values(list(df.columns)).head(20)

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
60,29,1,1,130,204,0,0,202,0,0.0,2,0,2,1
64,29,1,1,130,204,0,0,202,0,0.0,2,0,2,1
118,29,1,1,130,204,0,0,202,0,0.0,2,0,2,1
668,29,1,1,130,204,0,0,202,0,0.0,2,0,2,1
12,34,0,1,118,210,0,1,192,0,0.7,2,0,2,1
15,34,0,1,118,210,0,1,192,0,0.7,2,0,2,1
779,34,0,1,118,210,0,1,192,0,0.7,2,0,2,1
143,34,1,3,118,182,0,0,174,0,0.0,2,0,2,1
201,34,1,3,118,182,0,0,174,0,0.0,2,0,2,1
572,34,1,3,118,182,0,0,174,0,0.0,2,0,2,1


In [ ]:
df.nunique()

,0
age,41
sex,2
cp,4
trestbps,49
chol,152
fbs,2
restecg,3
thalach,91
exang,2
oldpeak,40


In [ ]:
df.drop_duplicates().shape

(302, 14)

In [ ]:
df.drop_duplicates()["target"].value_counts()

,count
target,
1,164
0,138


In [ ]:
df_clean = df.dropna()  # or df.copy(), depending on your cleaning step
X = df_clean.drop("target", axis=1)
y = df_clean["target"]

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test= train_test_split(
    X, y,test_size=0.2, random_state=42,stratify=y
)

In [ ]:
from sklearn.dummy import DummyClassifier

baseline = DummyClassifier(strategy="most_frequent")
baseline.fit(X_train, y_train)

baseline.score(X_test, y_test)

0.5121951219512195

In [ ]:
from sklearn.linear_model import LogisticRegression

log_reg = LogisticRegression(max_iter=1000)

log_reg.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [ ]:
y_pred = log_reg.predict(X_test)

In [ ]:
y_prob = log_reg.predict_proba(X_test)

y_prob[:5]

array([[0.98717103, 0.01282897],
       [0.61790684, 0.38209316],
       [0.99101024, 0.00898976],
       [0.27211006, 0.72788994],
       [0.70746934, 0.29253066]])

In [ ]:
from sklearn.metrics import accuracy_score

accuracy_score(y_test, y_pred)

0.8146341463414634

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [ ]:
from sklearn.pipeline import Pipeline

model = Pipeline([
    ("scaler", StandardScaler()),
    ("logistic", LogisticRegression(max_iter=1000))
])

model.fit(X_train, y_train)

Pipeline(steps=[('scaler', StandardScaler()),
                ('logistic', LogisticRegression(max_iter=1000))])

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train_scaled, y_train)
y_pred_knn=knn.predict(X_test_scaled)

In [ ]:
from sklearn.metrics import accuracy_score

accuracy_score(y_test, y_pred_knn)

0.8634146341463415

In [ ]:
for k in [1, 3, 5, 7, 9, 11, 15, 20]:
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train_scaled, y_train)

    y_pred_knn = knn.predict(X_test_scaled)

    print(k, accuracy_score(y_test, y_pred_knn))

1 1.0
3 0.9463414634146341
5 0.8634146341463415
7 0.8439024390243902
9 0.8536585365853658
11 0.8585365853658536
15 0.848780487804878
20 0.8439024390243902


In [ ]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

lda = LinearDiscriminantAnalysis()

lda.fit(X_train_scaled, y_train)

y_pred_lda = lda.predict(X_test_scaled)

accuracy_score(y_test, y_pred_lda)

0.8048780487804879

In [ ]:
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

qda = QuadraticDiscriminantAnalysis()

qda.fit(X_train_scaled, y_train)

y_pred_qda = qda.predict(X_test_scaled)

accuracy_score(y_test, y_pred_qda)

0.8292682926829268

In [ ]:
from sklearn.svm import SVC

svm = SVC(kernel="linear")

svm.fit(X_train_scaled, y_train)

y_pred_svm = svm.predict(X_test_scaled)

accuracy_score(y_test, y_pred_svm)

0.8146341463414634

In [ ]:
from sklearn.svm import SVC

svm_rbf = SVC(kernel="rbf")

svm_rbf.fit(X_train_scaled, y_train)

y_pred_svm = svm_rbf.predict(X_test_scaled)

accuracy_score(y_test, y_pred_svm)

0.926829268292683

In [ ]:
from sklearn.tree import DecisionTreeClassifier

tree = DecisionTreeClassifier(random_state=42)

tree.fit(X_train, y_train)

y_pred_tree = tree.predict(X_test)

accuracy_score(y_test, y_pred_tree)

0.9853658536585366

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

accuracy_score(y_test, y_pred_rf)

1.0

In [ ]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred_rf)

print(cm)

[[100   0]
 [  0 105]]


In [ ]:
from sklearn.model_selection import cross_val_score

scores = cross_val_score(
    rf,
    X_train,
    y_train,
    cv=5,
    scoring="accuracy"
)

print(scores)
print("Mean:", scores.mean())

[0.95121951 0.98780488 0.98780488 0.98170732 1.        ]
Mean: 0.9817073170731707


In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

log_reg_pipeline = Pipeline([
    ("scaler", StandardScaler()),
    ("model", LogisticRegression(max_iter=1000))
])

In [ ]:
models = {
    "Logistic Regression": log_reg_pipeline,
    "KNN": knn,
    "LDA": lda,
    "QDA": qda,
    "SVM": svm_rbf,
    "Decision Tree": tree,
    "Random Forest": rf
}

for name, model in models.items():
    scores = cross_val_score(
        model,
        X_train,
        y_train,
        cv=5,
        scoring="accuracy"
    )

    print(name, ":", round(scores.mean(), 4))

Logistic Regression : 0.8439
KNN : 0.7085
LDA : 0.8366
QDA : 0.8415
SVM : 0.7012
Decision Tree : 0.9866
Random Forest : 0.9817


In [ ]:
X = df_clean.drop("target", axis=1)
y = df_clean["target"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [ ]:
from sklearn.metrics import precision_score, recall_score

print("Precision:", precision_score(y_test, y_pred_rf))
print("Recall:", recall_score(y_test, y_pred_rf))

Precision: 1.0
Recall: 1.0


In [ ]:
from sklearn.metrics import f1_score

f1_score(y_test, y_pred_rf)

1.0

In [ ]:
from sklearn.metrics import roc_auc_score

y_prob_rf = rf.predict_proba(X_test)[:, 1]

auc = roc_auc_score(y_test, y_prob_rf)

print("AUC:", auc)

AUC: 1.0


In [ ]:
y_prob_rf = rf.predict_proba(X_test)[:, 1]

for threshold in [0.3, 0.4, 0.5, 0.6, 0.7]:
    y_pred_threshold = (y_prob_rf >= threshold).astype(int)

    print(
        threshold,
        "Precision:", precision_score(y_test, y_pred_threshold),
        "Recall:", recall_score(y_test, y_pred_threshold)
    )

0.3 Precision: 1.0 Recall: 1.0
0.4 Precision: 1.0 Recall: 1.0
0.5 Precision: 1.0 Recall: 1.0
0.6 Precision: 1.0 Recall: 1.0
0.7 Precision: 1.0 Recall: 1.0


## Fixing data leakage from duplicate rows

723 of the 1025 rows are exact duplicates, so the same patient can land in both the train and test sets. That is why Random Forest and the Decision Tree score 100% above. Below, everything is re-run on the 302 unique rows.

In [ ]:
df_unique = df.drop_duplicates().reset_index(drop=True)
X = df_unique.drop("target", axis=1)
y = df_unique["target"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)
print(X_train.shape, X_test.shape)

In [ ]:
from sklearn.pipeline import make_pipeline

models = {
    "Logistic Regression": make_pipeline(StandardScaler(), LogisticRegression(max_iter=1000)),
    "KNN (k=5)": make_pipeline(StandardScaler(), KNeighborsClassifier(n_neighbors=5)),
    "LDA": LinearDiscriminantAnalysis(),
    "QDA": QuadraticDiscriminantAnalysis(),
    "SVM (rbf)": make_pipeline(StandardScaler(), SVC(kernel="rbf", probability=True, random_state=42)),
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "Random Forest": RandomForestClassifier(n_estimators=100, random_state=42),
}

for name, m in models.items():
    scores = cross_val_score(m, X_train, y_train, cv=5, scoring="accuracy")
    print(f"{name:20s}: {scores.mean():.4f} ± {scores.std():.4f}")

## Confusion matrix with `cross_val_predict`

Predictions for every training row come from a model that never saw that row, so the test set stays untouched.

In [ ]:
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import ConfusionMatrixDisplay

clf = models["Logistic Regression"]
y_train_pred = cross_val_predict(clf, X_train, y_train, cv=5)

print(confusion_matrix(y_train, y_train_pred))
print("Precision:", precision_score(y_train, y_train_pred))
print("Recall:   ", recall_score(y_train, y_train_pred))
print("F1:       ", f1_score(y_train, y_train_pred))

ConfusionMatrixDisplay.from_predictions(y_train, y_train_pred)

## Precision/Recall trade-off

For a disease screen, missing a sick patient (false negative) is worse than a false alarm, so pick the threshold that gives at least 90% recall.

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import precision_recall_curve

y_scores = cross_val_predict(clf, X_train, y_train, cv=5, method="decision_function")
precisions, recalls, thresholds = precision_recall_curve(y_train, y_scores)

plt.figure(figsize=(8, 4))
plt.plot(thresholds, precisions[:-1], "b--", label="Precision")
plt.plot(thresholds, recalls[:-1], "g-", label="Recall")
plt.xlabel("Threshold")
plt.legend()
plt.grid(True)
plt.show()

In [ ]:
plt.figure(figsize=(5, 5))
plt.plot(recalls, precisions)
plt.xlabel("Recall")
plt.ylabel("Precision")
plt.title("Precision vs Recall")
plt.grid(True)
plt.show()

In [ ]:
idx = np.argmax(recalls[:-1] < 0.90) - 1
threshold_90_recall = thresholds[idx]
print("Threshold:", threshold_90_recall)

y_train_pred_90 = (y_scores >= threshold_90_recall)
print("Precision:", precision_score(y_train, y_train_pred_90))
print("Recall:   ", recall_score(y_train, y_train_pred_90))

## ROC curve: Logistic Regression vs Random Forest

In [ ]:
from sklearn.metrics import roc_curve

fpr, tpr, _ = roc_curve(y_train, y_scores)

y_probas_rf = cross_val_predict(models["Random Forest"], X_train, y_train, cv=5, method="predict_proba")
y_scores_rf = y_probas_rf[:, 1]
fpr_rf, tpr_rf, _ = roc_curve(y_train, y_scores_rf)

plt.figure(figsize=(6, 5))
plt.plot(fpr, tpr, "b:", label=f"Logistic Regression (AUC={roc_auc_score(y_train, y_scores):.3f})")
plt.plot(fpr_rf, tpr_rf, "g-", label=f"Random Forest (AUC={roc_auc_score(y_train, y_scores_rf):.3f})")
plt.plot([0, 1], [0, 1], "k--")
plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate (Recall)")
plt.legend()
plt.grid(True)
plt.show()

## Final evaluation on the test set

In [ ]:
from sklearn.metrics import classification_report

final_model = models["Logistic Regression"]
final_model.fit(X_train, y_train)

test_scores = final_model.decision_function(X_test)
y_test_pred = (test_scores >= threshold_90_recall).astype(int)

print(classification_report(y_test, y_test_pred))
print("ROC AUC:", roc_auc_score(y_test, test_scores))

## Multiclass classification

Predict chest-pain type `cp` (4 classes). SVC uses One-vs-One internally; `OneVsRestClassifier` forces One-vs-Rest.

In [ ]:
from sklearn.multiclass import OneVsRestClassifier

X_mc = df_unique.drop(["cp", "target"], axis=1)
y_mc = df_unique["cp"]
X_mc_train, X_mc_test, y_mc_train, y_mc_test = train_test_split(
    X_mc, y_mc, test_size=0.2, random_state=42, stratify=y_mc
)

ovo = make_pipeline(StandardScaler(), SVC(random_state=42))
ovr = make_pipeline(StandardScaler(), OneVsRestClassifier(SVC(random_state=42)))
sgd_mc = make_pipeline(StandardScaler(), LogisticRegression(max_iter=1000))

for name, m in [("SVC (OvO)", ovo), ("SVC (OvR)", ovr), ("Softmax LogReg", sgd_mc)]:
    print(f"{name:15s}:", cross_val_score(m, X_mc_train, y_mc_train, cv=3, scoring="accuracy").mean())

print("Baseline (most frequent):", y_mc_train.value_counts(normalize=True).max())

In [ ]:
y_mc_pred = cross_val_predict(sgd_mc, X_mc_train, y_mc_train, cv=3)
ConfusionMatrixDisplay.from_predictions(y_mc_train, y_mc_pred, normalize="true", values_format=".0%")
plt.title("Chest pain type – normalized confusion matrix")
plt.show()

## Multilabel classification

Predict two labels at once: heart disease (`target`) and exercise-induced angina (`exang`).

In [ ]:
X_ml = df_unique.drop(["target", "exang"], axis=1)
y_ml = df_unique[["target", "exang"]].values

X_ml_train, X_ml_test, y_ml_train, y_ml_test = train_test_split(
    X_ml, y_ml, test_size=0.2, random_state=42
)

knn_ml = make_pipeline(StandardScaler(), KNeighborsClassifier(n_neighbors=7))
knn_ml.fit(X_ml_train, y_ml_train)
print(knn_ml.predict(X_ml_test[:5]))

y_ml_pred = cross_val_predict(knn_ml, X_ml_train, y_ml_train, cv=3)
print("Macro F1:", f1_score(y_ml_train, y_ml_pred, average="macro"))